### PR TITLE
feat(osv): add osv support

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
Support OSV format
https://ossf.github.io/osv-schema/#python-vulnerability

## Blockers
- [x] https://github.com/aquasecurity/vuln-list-update/pull/113

### CODE REVIEW

File: pkg/vulnsrc/osv/osv.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	datasourceFormat = "osv-%s"
+	platformFormat   = "Osv Security Advisories %s"
+
+	//ecosystem names
+	Python = "PyPI"
+	Go     = "Go"
+	Rust   = "crates.io"
+)
+
+var defaultEcosystem = []ecosystem{
```
Comment (AndreyLevchenko) [inline pkg/vulnsrc/osv/osv.go:32]: consider using `map` here this way `getEcoSystem()` won't be required

Comment (AndreyLevchenko) [reply pkg/vulnsrc/osv/osv.go:32]: also variable name should be in plural form

File: pkg/vulnsrc/osv/osv.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+func NewVulnSrc(ecosystemName string) VulnSrc {
+	ecosystem := getEcoSystem(ecosystemName)
+	return VulnSrc{
+		ecosystem: ecosystem,
+		dbc:       db.Config{},
+	}
+}
+
+func (vs VulnSrc) Name() string {
+	switch vs.ecosystem.name {
```
Comment (AndreyLevchenko) [inline pkg/vulnsrc/osv/osv.go:59]: add `vulnerability.OsvPyPI`, `vulnerability.OsvGo` and `vulnerability.OsvCratesio` as separate property to `defaultEcosystems` map entries above. Get entry and return that property when name is required

**Review by AndreyLevchenko — CHANGES_REQUESTED** [review]

File: pkg/vulnsrc/osv/osv.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+func getEcoSystem(ecosystemName string) ecosystem {
+	for _, system := range defaultEcosystem {
+		if system.name == ecosystemName {
+			return system
+		}
+	}
+	return ecosystem{}
+}
+
+func (vs VulnSrc) Get(pkgName string) ([]types.Advisory, error) {
```
Comment (AndreyLevchenko) [inline pkg/vulnsrc/osv/osv.go:190]: this is very similar to
https://github.com/aquasecurity/trivy-db/blob/main/pkg/db/advisory.go#L23

Comment (DmitriyLewen) [reply pkg/vulnsrc/osv/osv.go:176]: used `GetAdvisories`

File: pkg/vulnsrc/osv/osv.go (hunk 2 above)
Comment (DmitriyLewen) [reply pkg/vulnsrc/osv/osv.go]: added new field in `ecosystem`, changed `Name()`

File: pkg/vulnsrc/osv/osv.go (hunk 1 above)
Comment (DmitriyLewen) [reply pkg/vulnsrc/osv/osv.go]: made `map`, remove `getEcoSystem()`

File: pkg/vulnsrc/osv/osv_test.go [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+						PublishedDate:    MustParse(time.RFC3339, "2018-07-12T12:29:00Z"),
+						LastModifiedDate: MustParse(time.RFC3339Nano, "2021-06-10T06:51:37.378319Z"),
+						Title:            "PYSEC-2018-27",
+						References:       []string{"https://github.com/qutebrowser/qutebrowser/commit/43e58ac865ff862c2008c510fc5f7627e10b4660"},
+					},
+				},
+			},
+		},
+		{
+			name:      "single range version and references Go",
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/vulnsrc/osv/osv_test.go]: changed test names

File: pkg/vulnsrc/osv/osv.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+func getVulnId(osv *vtypes.OsvJson) string {
+	if len(osv.Aliases) == 0 {
+		return osv.Id
+	} else {
+		return osv.Aliases[0] //CVE Id
+	}
+}
+
+func MustParse(layout, value string) *time.Time {
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/vulnsrc/osv/osv.go]: made it private

**Review by AndreyLevchenko — CHANGES_REQUESTED** [review]

File: pkg/vulnsrc/osv/osv.go (hunk 3 above)
Comment (DmitriyLewen) [reply pkg/vulnsrc/osv/osv.go]: removed `Get()`

File: pkg/vulnsrc/osv/osv.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+type ecosystem struct {
+	dir          string
+	dataSource   string
+	eventType    string
+	firstVersion string
+}
+
+type VulnSrc struct {
+	ecosystem ecosystem
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/vulnsrc/osv/osv.go]: done

File: pkg/vulnsrc/osv/osv.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	"github.com/aquasecurity/trivy-db/pkg/db"
+	"github.com/aquasecurity/trivy-db/pkg/types"
+	"github.com/aquasecurity/trivy-db/pkg/utils"
+	"github.com/aquasecurity/trivy-db/pkg/vulnsrc/vulnerability"
+	vtypes "github.com/aquasecurity/vuln-list-update/osv"
+)
+
+const (
+	osvDir         = "osv"
+	platformFormat = "%s::Osv Security Advisories"
```
Comment (knqyf263) [inline pkg/vulnsrc/osv/osv.go:22]: You can this function.
https://github.com/aquasecurity/trivy-db/blob/3c60651dc9c4195e57e7057b853a7f37859bf0fb/pkg/vulnsrc/bucket/bucket.go#L16

File: pkg/vulnsrc/osv/osv.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+	"time"
+
+	bolt "go.etcd.io/bbolt"
+	"golang.org/x/xerrors"
+
+	"github.com/aquasecurity/trivy-db/pkg/db"
+	"github.com/aquasecurity/trivy-db/pkg/types"
+	"github.com/aquasecurity/trivy-db/pkg/utils"
+	"github.com/aquasecurity/trivy-db/pkg/vulnsrc/vulnerability"
+	vtypes "github.com/aquasecurity/vuln-list-update/osv"
```
Comment (knqyf263) [inline pkg/vulnsrc/osv/osv.go:17]: Please don't depend on vuln-list-udpate. You can define a new struct having necessary fields only.

File: pkg/vulnsrc/osv/osv.go (hunk 7 above)
Comment (DmitriyLewen) [reply pkg/vulnsrc/osv/osv.go]: used this function

File: pkg/vulnsrc/osv/osv.go (hunk 8 above)
Comment (DmitriyLewen) [reply pkg/vulnsrc/osv/osv.go]: Done

**Comment by DmitriyLewen** [comment]
Hello @knqyf263 !
I fixed errors after merging main branch, can you check?

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**